messaging/src/main/java/org/axonframework/commandhandling/gateway/AbstractRetryScheduler.java

```diff
@@ -150,6 +193,42 @@ public B retryExecutor(ScheduledExecutorService retryExecutor) {
             return (B) this;
         }
 
+        /**
+         * Sets the {@link Predicate} used for testing whether failure is transient (returns {@code false}) or not (returns {@code true}).
+         * <p/>
+         * Non-transient failures are never retried.
+         * <p/>
+         * If configured, {@code nonTransientFailures} list is ignored.
+         *
+         * @param nonTransientFailurePredicate a {@link Predicate} (accepting a {@link Throwable} parameter) used for testing transiency of provided failure
+         * @return the current Builder instance, for fluent interfacing
+         */
+        public B nonTransientFailurePredicate(Predicate<Throwable> nonTransientFailurePredicate) {
```

Comment:
@sandjelkovic 

Can you clarify few things for me, please?

When combining predicates, I assume you think about the `or()` operator, right? If so, in what order should predicates be `or`-ed"?
I guess, default/existing predicate should go in the last place.

If my assumptions are correct, we can have two base methods in the builder: `nonTransientFailurePredicate()` and `addNonTransientFailurePredicate()`. Predicate set via `nonTransientFailurePredicate()` will replace any existing predicate, while `addNonTransientFailurePredicate()` will do `or`-ing supplied parameter predicate with existing one (i.e, `paramPredicate.or(existingPredicate)`).

Alternatively, we can maybe add `orBeforeNonTransientFailurePredicate()` and `orAfterNonTransientFailurePredicate()` instead of `addNonTransientFailurePredicate()` to support `or`-ing before and after for greater flexibility. We can even add `and`-ing methods with before and after order if necessary.

What do you prefer? 

Do you find this as complete, or will you like to have variants with lists of predicates? If so, how should these variants combine predicates from a supplied list?

FWIW, from my Axon framework user perspective, I'll probably go only with `nonTransientFailurePredicate()` as I  find combining predicates relatively easy. It looks more explicit to me if a user defines their own combination instead of relying on something provided by Axon, as this probably will always be less flexible, and users have to learn details of combinations implemented in those provided methods.

Tnx

Reply:
Hey @dmurat ,
> When combining predicates, I assume you think about the or() operator, right?

Correct, that was exactly what I had in mind

>  If so, in what order should predicates be or-ed"?

This shouldn't really matter as if any of the predicates return true, the exception would be considered non-transient. As opposed to regular interceptors that can modify the data, these predicates shouldn't really modify the data, that's why they are Predicates and not Functions. The `or` operation itself is also optimised so it will short-circuit if the first (left) element is true.

> Do you find this as complete, or will you like to have variants with lists of predicates? If so, how should these variants combine predicates from a supplied list?

To be honest, I prefer your suggestion with the `addNonTransientFailurePredicate()` method more than a method with a list, and the base `nonTransientFailurePredicate()` to be able to wipe/clean previous Predicate (chain)

> FWIW, from my Axon framework user perspective, I'll probably go only with nonTransientFailurePredicate() as I find combining predicates relatively easy. It looks more explicit to me if a user defines their own combination instead of relying on something provided by Axon, as this probably will always be less flexible, and users have to learn details of combinations implemented in those provided methods.

I agree this should definitely be the primary option, and in turn most flexible. It can be anything the user wants and it can wipe any previous configuration. However, I can also see a lot of value in providing an additional `add` method, if the user doesn't want to override all previous configuration but just wants to add a new filter to all existing ones. 
Say you are configuring the base list of predicates, but want an additional filter in case of some special circumstances. Maybe you want some additional filters in case the application is running in localhost or docker etc. In these cases, it is much easier to just add an additional predicate than to redefine the whole list which you might or might not have easy access to.

Reply:
Tnx, I believe I have all bits now 👍 . I will see what can I come up with 🙂